src/solver/mha/mha_ck_fa_v2_solver_forward.cpp

```diff
@@ -0,0 +1,268 @@
+    fmha_args.nhead_stride_q       = H_stride_q;
+    fmha_args.nhead_stride_k       = H_stride_k;
+    fmha_args.nhead_stride_v       = H_stride_v;
+    fmha_args.nhead_stride_o       = H_stride_o;
+    fmha_args.nhead_stride_lse     = S_q;
+    fmha_args.nhead_stride_bias    = 0;
+    fmha_args.nhead_stride_randval = S_q * S_k;
+    fmha_args.batch_stride_q       = N_stride_q;
+    fmha_args.batch_stride_k       = N_stride_k;
+    fmha_args.batch_stride_v       = N_stride_v;
+    fmha_args.batch_stride_o       = N_stride_o;
+    fmha_args.batch_stride_lse     = H_q * S_q;
+    fmha_args.batch_stride_bias    = 0;
+    fmha_args.batch_stride_randval = H_q * S_q * S_k;
+
+    // These are used for group mode, and we are in batch right now.
+    fmha_args.seqstart_q_ptr = nullptr;
+    fmha_args.seqstart_k_ptr = nullptr;
+
+    // Batch does not support padding, and we aren't using kvcache yet.
+    fmha_args.seqlen_k_ptr = nullptr;
+
+    fmha_args.s_randval         = false;
+    fmha_args.scale_s           = scale_s;
+    fmha_args.scale_p           = scale_p;
+    fmha_args.scale_o           = scale_o;
+    fmha_args.window_size_left  = 0;
+    fmha_args.window_size_right = 0;
+    fmha_args.mask_type         = static_cast<ck_tile::index_t>(fmha_traits.mask_type);
+
+    result.invoker_factory = [=](const std::vector<Kernel>&) {
+        return [=](const Handle& handle_, const AnyInvokeParams& raw_params) {
+            decltype(auto) params = raw_params.CastTo<miopen::mha::InvokeParams>();
+            const auto& dataFwd   = params.GetDataForward();
+
+            fmha_fwd_traits fmha_runtime_traits = fmha_traits;
+            fmha_fwd_args fmha_runtime_args     = fmha_args;
+
+            fmha_runtime_args.q_ptr        = dataFwd.qData;
+            fmha_runtime_args.k_ptr        = dataFwd.kData;
+            fmha_runtime_args.v_ptr        = dataFwd.vData;
+            fmha_runtime_args.o_ptr        = dataFwd.oData;
+            fmha_runtime_args.rand_val_ptr = nullptr;
+            fmha_runtime_args.bias_ptr     = nullptr;
+            fmha_runtime_args.lse_ptr      = nullptr;
+
+            // Top-left causal mask
+            if(dataFwd.mask == miopenMhaMask_t::miopenMhaMaskCausal)
+            {
+                fmha_runtime_traits.mask_type = mask_enum::mask_top_left;
+                fmha_runtime_args.mask_type =
+                    static_cast<ck_tile::index_t>(mask_enum::mask_top_left);
+                fmha_runtime_args.window_size_left  = -1;
+                fmha_runtime_args.window_size_right = 0;
+            }
+
+            fmha_runtime_traits.has_dropout = false;
+            float probability               = 0;
+
+            // TODO : Change API to take in probability value as host side value instead of device
+            // pointer to match CK API. Calling a blocking hipMemcpy will cause issues with stream,
+            // and isn't async.
+
+            fmha_runtime_args.p_drop = probability;
```

Comment:
Disabled probability for now with this PR. 
We can't do a blocking readback as it will use the wrong stream, and the async version won't allow us to set the value properly for the kernel invocation.

I think the best way to handle this is to take in a host value for probability in the future which will allow us to match without a need to readback.

Reply:
1) `hipMemcpy` should not be used directly, actually non-async version must net be used ever.
2) `himMemcpyAsync` with proper current stream from the handle can be used, but should not be used directly.
3) `Handle` must provide all the functions to readback the value in a stream-safe manner, but right now it does not, since it uses `hipMemcpy` under the hood.

A general rule is to hide all direct hip API calls in the `Handle`, but we violate this rule for some new specific functionality like events, of some other hip API functions. We can't wrap them all, but we've lost flexibility and portability and actually that rule can be revisit since OpenCL backend has gone and now everything works through hip runtime.

Partially I agree that an external API is broken and we must pass those values through host pointers/values, but the latest changes in the graph API actually support both ways and all the graph-api tensors have flag is the tensor devise or host, so it means eventually we have to support readback or explicitly check that flag.

Reply:
Oh good point about hiding the call behind handle as well.
That's likely something we will need to decide on in the future.

I think if it were host -> device, then it's fine, and `hipMemcpyAsync` would work, but device -> host requires synchronization, and in this case it's device -> host -> device, which ideally should be avoided. If it was just device -> host, then we could use callbacks, or events to synchronize properly without issue.